### PR TITLE
Weight support in RateLimiter

### PR DESCRIPTION
**Pull request by walec51**
Implementation of https://github.com/resilience4j/resilience4j/issues/642

### CODE REVIEW

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java [hunk 1]
```diff
@@ -144,13 +145,35 @@ public boolean acquirePermission() {
...
+    /**
+     * Reserving permissions is not supported in the spemaphore based
+     * implementation. Semaphores are totally blocking by it's nature. So this
+     * non-blocking API isn't supported. Use {@link #acquirePermission()}
+     *
+     * @throws UnsupportedOperationException always for this implementation
      */
     @Override
     public long reservePermission() {
-        return -1;
```
Comment (walec51) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: PS. I changed this behaviour to `throw new UnsupportedOperationException` because I think its more developer friendly. No one should call `reservePermissions` in a `SemaphoreBasedRateLimiter`. If some one is doing it he is not aware it is not supported and returning -1 will not explain that to him at first glance.

Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: This is a non-backward compatible change. But I like the idea to throw a `UnsupportedOperationException`. What do you think @storozhukBM

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java [hunk 2]
```diff
@@ -144,13 +145,35 @@ public boolean acquirePermission() {
...
-     * SemaphoreBasedRateLimiter is totally blocking by it's nature. So this non-blocking API isn't supported.
-     * It will return negative numbers all the time.
+     * Acquiring permissions of a specified weight is not supported in the
+     * spemaphore based implementation. Use {@link #acquirePermission()}
+     *
+     * @throws UnsupportedOperationException always for this implementation
+     */
+    @Override
+    public boolean acquirePermission(int weight) {
+        throw new UnsupportedOperationException("Acquiring permissions of a specified weight is not supported in the spemaphore based implementation");
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: We should be able to support it with a Semaphore as well.

Comment (RobWin) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: We can use `tryAcquire(int permits)` and `release(int permits)

Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: done

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java (hunk 1 above)
Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: I can revert this but I doubt that any properly working code is depending on the current behaviour :)

Comment (RobWin) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: I could live with the change.

**Comment by walec51** [comment]
one test from the `resilience4j-feign` module does not produce repeatable results,
it passed on my local machine but it failed on the CI server

**Comment by RobWin** [comment]
I will fix that.

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java (hunk 1 above)
Comment (storozhukBM) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: I think this case is already covered in our documentation `amount of nanoseconds you should wait for reserved permission. if negative, it means you failed to reserve`
I'm not against such change in principle, but throwing exception can potentially eliminate some JIT optimizations, so I'd ask @walec51 to perform and post benchmarks and check if there are any regressions.

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java [hunk 3]
```diff
@@ -201,25 +202,26 @@ private State calculateNextState(final long timeoutInNanos, final State activeSt
...
      * @param currentCycle         current {@link AtomicRateLimiter} cycle    @return nanoseconds to wait for the next permission
      */
-    private long nanosToWaitForPermission(final long cyclePeriodInNanos, final int permissionsPerCycle,
+    private long nanosToWaitForPermission(final int weight, final long cyclePeriodInNanos, final int permissionsPerCycle,
                                           final int availablePermissions, final long currentNanos, final long currentCycle) {
-        if (availablePermissions > 0) {
+        if (availablePermissions >= weight) {
             return 0L;
         }
         long nextCycleTimeInNanos = (currentCycle + 1) * cyclePeriodInNanos;
```
Comment (storozhukBM) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: Unfortunately, you can't just update a happy path for this function. With `weight` being introduced to this function, you will need to wait longer, potentially even more than one cycle to accumulate the required amount of permissions.

Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: I don't think that having weights > limitForPeriod is an actual use case that exists in the wild. One could argue if a `aquirePermission` in this situation should wait multiples of limitRefreshPeriod or should it always timeout in this situation. Having a `aquirePermission` that block even on a fresh cycle seams strange to me.

I would suggest to address this issue either by:
a) throwing an exception the passed weight is weights > limitForPeriod
b) adding some documentation that which ever weight you pass we will not wait more then one limitRefreshPeriod

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java (hunk 1 above)
Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: After reading the documentation I think most people will assume that `failed to reserve` means that it failed because the calculated wait time was more then the timeout. Calling an unsupported function is a different calibre of failure to me. The first type of failure is expected to happen in normal production code, the second is not.

As for testing performance you mean that I should test if the performance of calling `AtomicRateLimiter#reservePermission` if affected by this change?

Because as for `SemaphoreBasedRateLimiter#reservePermission` I do not think performance is any issue because it should never be called in production code and the whole point of throwing the suggested exception is to prevent that as early in development as possible.

Comment (RobWin) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: > After reading the documentation I think most people will assume that failed to reserve means that it failed because the calculated wait time was more then the timeout. Calling an unsupported function is a different calibre of failure to me. The first type of failure is expected to happen in normal production code, the second is not.

I agree

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java (hunk 3 above)
Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: so, in which direction do we want to take this?
should we have an `aquirePermission` that can wait even on a fresh cycle or should we do (a) or (b) from my proposals?
personally I'm in favour of (a)

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java (hunk 1 above)
Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: I tried running the JMH benchmarks but their Gradle configuration doesn't work on my machine.

```
cd resilience4j-ratelimiter
../gradlew --configure-on-demand -w -x check jmh
Configuration on demand is an incubating feature.
Processing 1 classes from /mnt/reszta/Projekty/WALCZAK IT/scr/resilience4j/resilience4j-ratelimiter/build/classes/java/jmh with "reflection" generator
Writing out Java source to /mnt/reszta/Projekty/WALCZAK IT/scr/resilience4j/resilience4j-ratelimiter/build/jmh-generated-sources and resources to /mnt/reszta/Projekty/WALCZAK IT/scr/resilience4j/resilience4j-ratelimiter/build/jmh-generated-resources
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.openjdk.jmh.util.Utils (file:/home/walec51/.gradle/caches/modules-2/files-2.1/org.openjdk.jmh/jmh-core/1.21/442447101f63074c61063858033fbfde8a076873/jmh-core-1.21.jar) to field java.io.PrintStream.charOut
WARNING: Please consider reporting this to the maintainers of org.openjdk.jmh.util.Utils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
# JMH version: 1.21
# VM version: JDK 11.0.4, OpenJDK 64-Bit Server VM, 11.0.4+11
# VM invoker: /usr/lib/jvm/java-11-openjdk-11.0.4.11-0.fc30.x86_64/bin/java
# VM options: <none>
# Warmup: 10 iterations, 10 s each
# Measurement: 10 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 2 threads, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.github.resilience4j.ratelimiter.RateLimiterBenchmark.atomicPermission

# Run progress: 0,00% complete, ETA 00:40:00
# Fork: 1 of 2
Error: Could not find or load main class org.openjdk.jmh.runner.ForkedMain
Caused by: java.lang.ClassNotFoundException: org.openjdk.jmh.runner.ForkedMain
<forked VM failed with exit code 1>
<stdout last='20 lines'>
</stdout>
<stderr last='20 lines'>
Error: Could not find or load main class org.openjdk.jmh.runner.ForkedMain
Caused by: java.lang.ClassNotFoundException: org.openjdk.jmh.runner.ForkedMain
</stderr>
```

any suggestions what could be the cause of this?

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java (hunk 3 above)
Comment (storozhukBM) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: I believe that we should not change the contract and implement current semantic "if an `aquirePermission` in this situation should wait for a multiple of `limitRefreshPeriod's`". The client already can pass the expected timeout, so if `nanosToWaitForPermission` returns too big duration, we won't reserve these permissions.

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java (hunk 1 above)
Comment (RobWin) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java]: I have a look at it. We did a major gradle update a while ago.

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java (hunk 3 above)
Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: so...
even if the rate limiter did do nothing for the past dozens of cycles,
if I call `aquirePermission(6)` and my `limitForPeriod = 2`,
then you want me to wait for 3 cycles?

this behaviour is absurd but if you want it I'll implement it
from my perspective its an error to have weight > limitForPeriod and I would never allow it in my application code nor does any API service known to me impose such bizarre rules - they would be even impossible to enforce from the APIs perspective because you'll wait even before making your first call to a function with a weight > limitForPeriod ...

Comment (storozhukBM) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: Hi @walec51

First of all, I want to say that I'm very happy and thankful that you decided to contribute such a change to our library. I think that this is a good idea and it will be valuable for some of our users.
But I want to kindly ask you to try to avoid any subjective thoughts and messages. Things like "so..." and "this behavior is absurd" can be considered rude by other contributors.

As for your point about, "its an error to have `weight` > `limitForPeriod` and I would never allow it in my application code". I totally get it, especially if we consider only "blocking" APIs.
But current AtomicRateLimiter is essentially an implementation of a super simple "fixed window rate limiting" algorithm with non-blocking API. And in our implementation, it is OK for `State.activePermissions` to be a negative `int`, where it means that there is a thread or multiple threads that reserved some permissions for future calls and are waiting for their turn or doing some other work while waiting. 
So when we take this into consideration, there can be a situation where
`weight` is less than `limitForPeriod`, but you'll need to wait for more than one cycle for your turn to perform a request.

Hopefully, this makes sense for you, and please correct me if I'm wrong.

Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: ok, sorry for my attitude, please take a look at the current implementation

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java [hunk 4]
```diff
@@ -201,50 +202,56 @@ private State calculateNextState(final long timeoutInNanos, final State activeSt
...
         long nanosToNextCycle = nextCycleTimeInNanos - currentNanos;
-        int fullCyclesToWait = (-availablePermissions) / permissionsPerCycle;
-        return (fullCyclesToWait * cyclePeriodInNanos) + nanosToNextCycle;
+        int permissionsToTakeAfterCurrencyCycle = weight - availablePermissions;
+        int fullCyclesToWaitCount = Math.floorDiv(permissionsToTakeAfterCurrencyCycle, permissionsPerCycle);
+        if (permissionsToTakeAfterCurrencyCycle % permissionsPerCycle == 0) {
+            // will take all permissions from the last cycle but can reserve them without waiting till its end
+            fullCyclesToWaitCount -= 1;
+        }
+        return nanosToNextCycle + (fullCyclesToWaitCount * cyclePeriodInNanos);
```
Comment (storozhukBM) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: I'd suggest such simplification
```java
int permissionsAtTheStartOfNextCycle = availablePermissions + permissionsPerCycle;
int fullCyclesToWait = -(permissionsAtTheStartOfNextCycle-weight) / permissionsPerCycle;
return (fullCyclesToWait * cyclePeriodInNanos) + nanosToNextCycle;
```
I'm not totally sure about all corner cases, but all tests appear to work OK.
If I missed something, please correct me.

Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java]: ok, done

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiter.java [hunk 5]
```diff
@@ -318,7 +318,23 @@ static void waitForPermission(final RateLimiter rateLimiter) {
...
+	 *
+	 * <p>If the current thread is {@linkplain Thread#interrupt interrupted}
+	 * while waiting for a permit then it won't throw {@linkplain InterruptedException},
+	 * but its interrupt status will be set.
+	 * 
+	 * @param weight weight of permission - use for systems where 1 call != 1 permission
+	 * @return {@code true} if a permit was acquired and {@code false}
+	 * if waiting timeoutDuration elapsed before a permit was acquired
+	 */
+	boolean acquirePermission(int weight);
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiter.java]: Can we rename `weight` to `permits`. Analogous to `https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html#acquire(int)`?

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiter.java [hunk 6]
```diff
@@ -327,7 +343,19 @@ static void waitForPermission(final RateLimiter rateLimiter) {
...
+
+	/**
+	 * Reserves a permission of a given weight from this rate limiter and returns nanoseconds you should wait for it.
+	 * If returned long is negative, it means that you failed to reserve permission,
+	 * possibly your  {@link RateLimiterConfig#getTimeoutDuration()} is less then time to wait for permission.
+	 *
+	 * @param weight weight of permission - use for systems where 1 call != 1 permission
+	 * @return {@code long} amount of nanoseconds you should wait for reserved permission. if negative, it means you failed to reserve.
+	 */
+	long reservePermission(int weight);
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiter.java]: Can we rename `weight` to `permits`. Analogous to `https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html#acquire(int)`?

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/AbstractRateLimiterEvent.java [hunk 7]
```diff
@@ -35,6 +41,11 @@ public String getRateLimiterName() {
         return rateLimiterName;
     }
 
+    @Override
+    public int getWeight() {
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/AbstractRateLimiterEvent.java]: Can we rename it to `getNumberOfPermits`?

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/AbstractRateLimiterEvent.java [hunk 8]
```diff
@@ -23,10 +23,16 @@
...
 
     private final String rateLimiterName;
+    private final int weight;
     private final ZonedDateTime creationTime;
 
     public AbstractRateLimiterEvent(String rateLimiterName) {
+        this(rateLimiterName, 1);
+    }
+
+    public AbstractRateLimiterEvent(String rateLimiterName, int weight) {
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/AbstractRateLimiterEvent.java]: Can we rename `weight` to `permits`.

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterEvent.java [hunk 9]
```diff
@@ -28,6 +28,8 @@
 
     Type getEventType();
 
+    int getWeight();
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterEvent.java]: Can we rename it to getNumberOfPermits?

**Comment by RobWin** [comment]
PR looks good to me.
@storozhukBM Your final review?

**Comment by storozhukBM** [comment]
Hi @RobWin and @walec51 
Looks good for me too. So, I'll proceed with the merge. 
@walec51 thank you once again for this contribution and your patience, good job.
